exports/dataset.cpp

```diff
@@ -56,7 +70,24 @@ template <class T> void declare_span(py::module &m, const std::string &suffix) {
       .def("__len__", &gsl::span<T>::size)
       .def("__iter__", [](const gsl::span<T> &self) {
         return py::make_iterator(self.begin(), self.end());
-      });
+      })
+      .def("__repr__",
+       [](const gsl::span<T> &self) {
+         const gsl::index size = self.size();
+         if (size == 0)
+           return std::string("[]");
+         std::string s = "[";
+         for (gsl::index i = 0; i < self.size(); ++i) {
+           if (i == 4 && size > 8) {
+             s += "..., ";
+             i = size - 4;
+           }
+           s += print(self[i]);
```

Comment:
`print`? I don't think this function is doing any input/output. Maybe `to_string` is better and more consistent.

Reply:
@OwenArnold, print calls to_string with the same signature, so it wouldn't work, although I agree print is not perfect here.

Reply:
something inside the `print()` function is already called `to_string`. Any ideas for another name?

Reply:
I have 2 suggestions:
1.
``` c++
namespace tostr {
using namespace std::to_string;
using namespace dataset::to_string;
template <class T>
std::string impl(const T& x) { return to_string(x); }
} // tostr

namespace detail {
template <class T> std::string to_string(const T &item) {
  if constexpr (std::is_same_v<T, std::string>)
    return {'"' + item + "\", "};
  else if constexpr (std::is_same_v<T, Eigen::Vector3d>)
    return {"(Eigen::Vector3d), "};
  else if constexpr (std::is_same_v<T,
                                    boost::container::small_vector<double, 8>>)
    return {"(vector), "};
  else
    return tostr::impl(item) + ", ";
}
} // detail
```

2.

```c++
emplate<typename T> struct HasStdToString{
private:
    static void detect(...);
    template<typename U> static decltype(std::to_string(std::declval<U>())) detect(const U&);
public:
    static constexpr bool value = !std::is_same<void, decltype(detect(std::declval<T>()))>::value;  // 
};

template <class T> std::string to_string(const T &item) {
  if constexpr (std::is_same_v<T, std::string>)
    return {'"' + item + "\", "};
  else if constexpr (std::is_same_v<T, Eigen::Vector3d>)
    return {"(Eigen::Vector3d), "};
  else if constexpr (std::is_same_v<T,
                                    boost::container::small_vector<double, 8>>)
    return {"(vector), "};
  else if constexpr (HasStdToString)
    return std::to_string(item) + ", ";
  else
    return dataset::to_string(item) + ", "';
}
```